cmd/cstorvolumeclaim/cstorvolumeclaim.go

```diff
@@ -0,0 +1,450 @@
+	sc *storagev1.StorageClass,
+) string {
+
+	spcName := sc.Parameters["storagePoolClaim"]
+	return spcName
+}
+
+// listCStorPools get the list of available pool using the storagePoolClaim
+// as labelSelector.
+func listCStorPools(
+	spcName string,
+	replicaCount int,
+) (*apis.CStorPoolList, error) {
+
+	if spcName == "" {
+		return nil, errors.New("failed to list cstorpool: spc name missing")
+	}
+
+	labelSelector := spcAnnotation + spcName
+
+	cstorPoolList, err := csp.KubeClient().List(metav1.ListOptions{
+		LabelSelector: labelSelector,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to list cstorpool for spc {%s}",
+			spcName,
+		)
+	}
+	if len(cstorPoolList.Items) < replicaCount {
+		return nil, errors.New("not enough pools available to create replicas")
+	}
+	return cstorPoolList, nil
+}
+
+// getOrCreateTargetService creates cstor volume target service
+func getOrCreateTargetService(storageClassName string,
+	claim *apis.CStorVolumeClaim,
+) (*corev1.Service, error) {
+
+	svcObj, err := svc.NewKubeClient(svc.WithNamespace("openebs")).
+		Get(claim.Name, metav1.GetOptions{})
+
+	if err == nil {
+		return svcObj, nil
+	}
+
+	// error other than 'not found', return err
+	if !k8serror.IsNotFound(err) {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get cstorvolume service {%v}",
+			svcObj.Name,
+		)
+	}
+
+	// Not found case, so need to create
+	svcObj, err = svc.NewBuilder().
+		WithName(claim.Name).
+		WithLabelsNew(getTargetServiceLabels(claim)).
+		WithOwnerReferenceNew(getTargetServiceOwnerReference(claim)).
+		WithSelectorsNew(getTargetServiceSelectors(claim)).
+		WithPorts(cvPorts).
+		Build()
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to build target service {%v}",
+			svcObj,
+		)
+	}
+
+	svcObj, err = svc.NewKubeClient(svc.WithNamespace("openebs")).Create(svcObj)
+	return svcObj, err
+}
+
+// getOrCreateCStorVolumeResource creates CStorVolume resource for a cstor volume
+func getOrCreateCStorVolumeResource(
+	service *corev1.Service,
+	claim *apis.CStorVolumeClaim,
+	class *storagev1.StorageClass,
+) (*apis.CStorVolume, error) {
+
+	qCap := claim.Spec.Capacity[corev1.ResourceStorage]
+	rfactor, err := getReplicationFactor(class)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get replica-count and factor from sc {%s}",
+			class.Name,
+		)
+	}
+
+	cfactor := rfactor/2 + 1
+
+	cvObj, err := cv.NewKubeclient(cv.WithNamespace("openebs")).
+		Get(claim.Name, metav1.GetOptions{})
+	if err != nil && !k8serror.IsNotFound(err) {
+		return nil, errors.Wrapf(
+			err,
+			"failed to get cstorvolume {%v}",
+			cvObj.Name,
+		)
+	}
+	if k8serror.IsNotFound(err) {
+		cvObj, err = cv.NewBuilder().
+			WithName(claim.Name).
+			WithLabelsNew(getCVLabels(claim)).
+			WithOwnerRefernceNew(getCVOwnerReference(claim)).
+			WithTargetIP(service.Spec.ClusterIP).
+			WithCapacity(qCap.String()).
+			WithCStorIQN(claim.Name).
+			WithNodeBase(cv.CStorNodeBase).
```

Comment:
What is a NodeBase?

Reply:
Its just the prefix base name of iqn `"iqn.2016-09.com.openebs.cstor"` which we are populating in CV resources